astropy/units/core.py

```diff
@@ -2134,20 +2118,17 @@ def __call__(
                     warnings.warn(msg, UnitsWarning)
                 return UnrecognizedUnit(s)
 
-        # can't use units.typing.Complex because of an import loop
-        elif isinstance(s, (int, float, complex, Fraction, np.number)):
+        from .typing import Complex
```

Comment:
Do add a comment that this `isinstance` check will cover all the ones explicitly listed before -- I certainly had to do a double-take seeing this, and we shouldn't have to! 

I also wonder if we should not just define our own `Number` (or use `numbers.Number` after all...) - that way we would not need a comment.

Note that this more than doubles the time for the most common cases (`int`, `float`). Though since the whole initialization time is 30 times longer and initializing with a number is rare, it arguably doesn't matter much.

Reply:
> Do add a comment that this `isinstance` check will cover all the ones explicitly listed before -- I certainly had to do a double-take seeing this, and we shouldn't have to!

It is very simple to look up the definition of `Complex`. What do you need a comment for? Would it be less confusing if it the alias was named `ComplexNumber`?

> I also wonder if we should not just define our own `Number` (or use `numbers.Number` after all...) - that way we would not need a comment.

The standard library `numbers` predates and is incompatible with typing.

> Note that this more than doubles the time for the most common cases (`int`, `float`). Though since the whole initialization time is 30 times longer and initializing with a number is rare, it arguably doesn't matter much.

Whoever wants a performant way of creating dimensionless units should use `CompositeUnit` directly.